crates/ruff_python_syntax_errors/src/lib.rs

```diff
@@ -105,6 +119,35 @@ impl SyntaxChecker {
                 });
             }
         }
+
+        self.duplicate_type_parameter_name(stmt);
+    }
+
+    fn duplicate_type_parameter_name(&mut self, stmt: &ast::Stmt) {
+        let (Stmt::FunctionDef(ast::StmtFunctionDef { type_params, .. })
+        | Stmt::ClassDef(ast::StmtClassDef { type_params, .. })
+        | Stmt::TypeAlias(ast::StmtTypeAlias { type_params, .. })) = stmt
+        else {
+            return;
+        };
+
+        let Some(type_params) = type_params else {
+            return;
+        };
+
+        for type_param in type_params
+            .iter()
+            .rev()
+            .sorted_by_key(|t| &t.name().id)
```

Comment:
Should we return early if `type_params.len < 2` to avoid the "hidden" allocation in `sorted_by_key`. That should allow us to skip this check for most type params.